clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java

```diff
@@ -0,0 +1,193 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Abstract assignor implementation which uses the following metadata format:
+ *
+ * ProtocolMetadata => Topics TopicPattern MetadataHash
+ *   Topics         => [String]
+ *   TopicPattern   => String
+ *   MetadataHash   => Bytes
+ *
+ * This class handles metadata consistency checking between group members. Custom assignors will
+ * usually extend this class unless they need to provide their own metadata format.
+ */
+public abstract class AbstractPartitionAssignor implements PartitionAssignor<AbstractPartitionAssignor.ConsumerMetadata> {
+    private static final String TOPIC_KEY_NAME = "topics";
+    private static final String TOPIC_PATTERN_KEY_NAME = "topic_pattern";
+    private static final String METADATA_HASH_KEY_NAME = "metadata_hash";
+
+    public static final Schema CONSUMER_METADATA_SCHEMA = new Schema(
+            new Field(TOPIC_KEY_NAME, new ArrayOf(Type.STRING)),
+            new Field(TOPIC_PATTERN_KEY_NAME, Type.STRING),
+            new Field(METADATA_HASH_KEY_NAME, Type.BYTES));
+
+    private static final Type CONSUMER_METADATA_TYPE = new ConsumerMetadataType();
+
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    public abstract List<TopicPartition> assign(String consumerId,
+                                                Map<String, Integer> partitionsPerTopic,
+                                                Map<String, ConsumerMetadata> consumerMetadata);
+
+    public Type schema() {
+        return CONSUMER_METADATA_TYPE;
+    }
+
+    @Override
+    public AssignmentResult assign(String consumerId,
+                                   Map<String, ConsumerMetadata> consumers,
+                                   MetadataSnapshot metadataSnapshot) {
+        byte[] topicMetadataHash = metadataSnapshot.hash();
+        Set<String> allSubscribedTopics = new HashSet<>();
+        boolean consistentMetadata = true;
+
+        for (ConsumerMetadata metadata : consumers.values()) {
+            byte[] memberTopicMetadataHash = Utils.toArray(metadata.topicMetadataHash);
+            consistentMetadata = consistentMetadata && Arrays.equals(topicMetadataHash, memberTopicMetadataHash);
+            allSubscribedTopics.addAll(metadata.topics);
+        }
+
+        if (!consistentMetadata) {
+            log.debug("Assignment failed due to inconsistent metadata among consumers");
+            return AssignmentResult.inconsistentMetadata(allSubscribedTopics);
+        }
+
+        Map<String, Integer> partitionsPerTopic = new HashMap<>();
+        for (String topic : allSubscribedTopics) {
+            Integer numPartitions = metadataSnapshot.topicMetadata().get(topic);
+            if (numPartitions != null)
+                partitionsPerTopic.put(topic, numPartitions);
+            else
+                log.debug("Skipping assignment for topic {} since no metadata is available", topic);
```

Comment:
The case I was trying to handle is non-existing topics. The current behavior of the consumer is to keep trying to fetch metadata for the non-existing topic in the hope that it will eventually be created, and to continue consuming from other subscribed topics. If one of the consumers does have metadata (perhaps because the topic was just created), then the metadata hash will be inconsistent and members will refetch. If the topic is later created, we will discover it when we update metadata and the consumers will rejoin.